src/couch_stats/src/csrt_query.erl

```diff
@@ -0,0 +1,164 @@
+    find_by_pid/1,
+    find_by_pidref/1,
+    find_workers_by_pidref/1,
+    group_by/2,
+    group_by/3,
+    sorted/1,
+    sorted_by/1,
+    sorted_by/2,
+    sorted_by/3
+]).
+
+%%
+%% Aggregate query API
+%%
+
+active() ->
+    active_int(all).
+
+active_coordinators() ->
+    active_int(coordinators).
+
+active_workers() ->
+    active_int(workers).
+
+%% active_json() or active(json)?
+active(json) ->
+    to_json_list(active_int(all)).
+
+active_coordinators(json) ->
+    to_json_list(active_int(coordinators)).
+
+active_workers(json) ->
+    to_json_list(active_int(workers)).
+
+active_int(coordinators) ->
+    select_by_type(coordinators);
+active_int(workers) ->
+    select_by_type(workers);
+active_int(all) ->
+    select_by_type(all).
+
+select_by_type(coordinators) ->
+    ets:select(?CSRT_ETS, ets:fun2ms(fun(#rctx{type = #coordinator{}} = R) -> R end));
+select_by_type(workers) ->
+    ets:select(?CSRT_ETS, ets:fun2ms(fun(#rctx{type = #rpc_worker{}} = R) -> R end));
+select_by_type(all) ->
+    ets:tab2list(?CSRT_ETS).
+
+find_by_nonce(Nonce) ->
+    csrt_server:match_resource(#rctx{nonce = Nonce}).
+
+find_by_pid(Pid) ->
+    csrt_server:match_resource(#rctx{pid_ref = {Pid, '_'}}).
+
+find_by_pidref(PidRef) ->
+    csrt_server:match_resource(#rctx{pid_ref = PidRef}).
+
+find_workers_by_pidref(PidRef) ->
+    csrt_server:match_resource(#rctx{type = #rpc_worker{from = PidRef}}).
+
+field(#rctx{pid_ref = Val}, pid_ref) -> Val;
+%% NOTE: Pros and cons to doing these convert functions here
+%% Ideally, this would be done later so as to prefer the core data structures
+%% as long as possible, but we currently need the output of this function to
+%% be jiffy:encode'able. The tricky bit is dynamically encoding the group_by
+%% structure provided by the caller of *_by aggregator functions below.
+%% For now, we just always return jiffy:encode'able data types.
+field(#rctx{nonce = Val}, nonce) -> Val;
+field(#rctx{type = Val}, type) -> csrt_util:convert_type(Val);
+field(#rctx{dbname = Val}, dbname) -> Val;
+field(#rctx{username = Val}, username) -> Val;
+field(#rctx{db_open = Val}, db_open) -> Val;
+field(#rctx{docs_read = Val}, docs_read) -> Val;
+field(#rctx{docs_written = Val}, docs_written) -> Val;
+field(#rctx{rows_read = Val}, rows_read) -> Val;
+field(#rctx{changes_returned = Val}, changes_returned) -> Val;
+field(#rctx{ioq_calls = Val}, ioq_calls) -> Val;
+field(#rctx{js_filter = Val}, js_filter) -> Val;
+field(#rctx{js_filtered_docs = Val}, js_filtered_docs) -> Val;
+field(#rctx{get_kv_node = Val}, get_kv_node) -> Val;
+field(#rctx{get_kp_node = Val}, get_kp_node) -> Val;
+field(#rctx{started_at = Val}, started_at) -> Val;
+field(#rctx{updated_at = Val}, updated_at) -> Val.
+
+curry_field(Field) ->
+    fun(Ele) -> field(Ele, Field) end.
+
+count_by(KeyFun) ->
+    group_by(KeyFun, fun(_) -> 1 end).
+
+group_by(KeyFun, ValFun) ->
+    group_by(KeyFun, ValFun, fun erlang:'+'/2).
+
+%% eg: group_by(mfa, docs_read).
```

Comment:
The `mfa` field is no longer present in `#rctx` record. Please remove it from the examples. Since the example in the provided form do not work.